### PR TITLE
add rdp simplify

### LINKED ISSUE #117: simplifying lines using Douglas–Peucker algorithm
sample implementation: 

https://github.com/paulmach/orb/blob/master/simplify/douglas_peucker.go
https://github.com/deet/simpleline/blob/master/rdp.go
https://gist.github.com/neilvallon/02cc9f6f31d10d72823c#file-rdp-go

### PR DESCRIPTION
**Pull request by jackielii**
fixes #117

### CODE REVIEW

File: xy/rdp_simplify.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
+package xy
+
+import (
+	"math"
+)
+
+const stride = 2
```
Comment (twpayne) [inline xy/rdp_simplify.go]: This should be a parameter, not a fixed constant.

File: xy/rdp_simplify.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+const (
+	earthRadius = 6371000
+)
+
+// MeterToAngleThreshold is a helper function to convert meter to lat lon delta threshold
+func MeterToAngleThreshold(meter float64) float64 {
+	a := math.Atan(meter / earthRadius)
+	return a * 180 / math.Pi
+}
```
Comment (twpayne) [inline xy/rdp_simplify.go]: `earthRadius` and `MeterToAngleThreshold` should be removed - the `xy` package is about planar geometry and has nothing to do with spheres.

File: xy/rdp_simplify.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+// SimplifyIndexes uses the Douglas-Peucker simplify method to simplify a 2D
+// flatCoords.
+//
+// Threshold is the distance between a point and the selected start
+// and end line segment. It returns the indexes of the points.
+//
+// Note that the indexes are based on points, So acesss to x, y pair should be:
+//  x:=flatCoords[i*stride]
+//  y:=flatCoords[i*stride+1]
+func SimplifyIndexes(flatCoords []float64, threshold float64) []int {
```
Comment (twpayne) [inline xy/rdp_simplify.go]: This function should be renamed to `SimplifyFlatCoords` and take `stride` as a parameter.

File: xy/rdp_simplify_test.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+			[]float64{0, 0, 0, 1, -1, 2, 0, 3, 0, 4, 1, 4, 2, 4.5, 3, 4, 3.5, 4, 4, 4},
+			0.4999,
+			[]int{0, 2, 4, 6, 9},
+		},
+		{
+			// https://github.com/golang/geo/issues/48
+			[]float64{0, 0, 0, 1, -1, 2, 0, 3, 0, 4, 1, 4, 2, 4.5, 3, 4, 3.5, 4, 4, 4},
+			0.5,
+			[]int{0, 2, 4, 9},
+		},
```
Comment (twpayne) [inline xy/rdp_simplify_test.go]: Please name struct fields explicitly.

File: xy/rdp_simplify.go (hunk 1 above)
Comment (jackielii) [reply xy/rdp_simplify.go]: Hi, thanks for reviewing.

I had a geom.Layout parameter, as you can see below all using stride variables. But the package already implies it's 2D geometry. What do you think?

File: xy/rdp_simplify.go (hunk 2 above)
Comment (jackielii) [reply xy/rdp_simplify.go]: Agreed, I just though that the threshold is a bit hard to understand as input, especially when the coordinates are in lat/lons. Will remove

File: xy/rdp_simplify.go (hunk 3 above)
Comment (jackielii) [reply xy/rdp_simplify.go]: `func SimplifyIndexes(stride int, flatCoords []float64, threshold float64) []int`
or 
`func SimplifyIndexes(layout geom.Layout, flatCoords []float64, threshold float64) []int`

Comment (twpayne) [reply xy/rdp_simplify.go]: See https://github.com/twpayne/go-geom/blob/master/flat.go#L351 for example.

**Comment by jackielii** [comment]
@twpayne all done, please can you review again? thanks

**Comment by twpayne** [comment]
Nice contribution, thank you, and thanks for your patience in the review :)

**Comment by jackielii** [comment]
Thanks for merging!

**Pull request merged**